state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1215,154 +1209,241 @@ pub mod pallet {
+					if is_gas_swap {
+						*gas_swap_state = GasSwapState::OutputReady { gas_budget: output_amount };
+					} else {
+						// The executed swap must be for the principal amount:
+						dca_state.accumulated_output_amount += output_amount;
+
+						// Schedule any remaining DCA chunks:
+						dca_state.scheduled_chunk_swap_id = if dca_state.remaining_chunks > 0 {
+							let chunk_input_amount = dca_state.prepare_next_chunk();
+
+							let swap_id = Self::schedule_swap(
+								request.input_asset,
+								request.output_asset,
+								chunk_input_amount,
+								swap.swap.refund_params.clone(),
+								SwapType::CcmPrincipal,
+								swap_request_id,
 							);
+
+							Some(swap_id)
 						} else {
-							log_or_panic!(
-								"IngressEgressFee swap should not be to non-gas asset: {:?}",
-								swap.output_asset()
+							None
+						};
+
+						// See if we still need to schedule the gas swap
+						if let GasSwapState::ToBeScheduled { gas_budget, other_gas_asset } =
+							gas_swap_state
+						{
+							*gas_swap_state = Self::schedule_ccm_gas_swap(
+								swap_request_id,
+								request.input_asset,
+								*other_gas_asset,
+								*gas_budget,
 							);
 						}
-					},
-				}
+					}
+
+					// Process outcome if there aren't any more scheduled swaps,
+					// otherwise put the request state back:
+					if let (GasSwapState::OutputReady { gas_budget }, None) =
+						(gas_swap_state, dca_state.scheduled_chunk_swap_id)
+					{
+						Self::process_ccm_outcome(
+							swap_request_id,
+							request.output_asset,
+							output_address.clone(),
+							dca_state.accumulated_output_amount,
+							*gas_budget,
+							ccm_deposit_metadata.clone(),
+						);
+					} else {
+						SwapRequests::<T>::insert(swap_request_id, request);
+					}
+				},
+				SwapRequestState::Regular { output_address, dca_state } => {
+					if dca_state.remaining_chunks > 0 {
+						let chunk_input_amount = dca_state.prepare_next_chunk();
+						dca_state.accumulated_output_amount += output_amount;
+
+						Self::schedule_swap(
+							request.input_asset,
+							request.output_asset,
+							chunk_input_amount,
+							// TODO: recompute swap params depending on the output amount we got
+							swap.swap.refund_params.clone(),
+							SwapType::Swap,
+							request.id,
+						);
+
+						// Not done with the request, put it back
+						SwapRequests::<T>::insert(swap_request_id, request);
+					} else {
+						debug_assert!(dca_state.remaining_input_amount == 0);
+
+						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
+
+						Self::egress_for_swap(
+							swap_request_id,
+							output_amount + dca_state.accumulated_output_amount,
+							swap.output_asset(),
+							output_address.clone(),
+							None,  /* ccm */
+							false, /* refund */
+						);
+					}
+				},
+				SwapRequestState::NetworkFee => {
+					Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
```

Comment:
It feels like there ought to be a way to deduplicate this... It's not easy to see if this event is being emitted in all the right places (ie. that every request eventually completes).

As a step in that direction, we could require that this match statement returns some indication of whether the swap completed or not. Then at least we can be sure that each match arm handles the question of whether the swap is completed or not.

Reply:
> we could require that this match statement returns some indication of whether the swap completed or not.

Updated to do this. As a bonus, now putting the request back into storage is done in one place also.

Reply:
minor nit: an enum would have been nicer than a bool. Using bools, when reading the code and seeing that some match arm returns true/false, you need to scroll up again to remind yourself what the true/false means. Something like RequestStatus::Complete, for example, would be friendlier.